src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,769 @@
+                SelectClauseSyntax selectClause,
+                QueryContinuationSyntax queryContinuation,
+                ProcessingState processingState)
+            {
+                // TODO semantic check for selectClause.Expression
+                // End of the queue
+                if (queryContinuation == null)
+                {
+                    processingState.AddStatementProcessingMethod(
+                        s => WrapWithBlock(processingState.LeafExpressionCreationMethod(selectClause.Expression)));
+                    return true;
+                }
+                else
+                {
+                    SymbolInfo symbolInfo = _semanticModel.GetSymbolInfo(queryContinuation, _cancellationToken);
+                    if (symbolInfo.Symbol == null ||
+                        symbolInfo.Symbol is IMethodSymbol methodSymbol &&
+                        methodSymbol.Parameters.Length == 1)
+                    {
+                        processingState.AddVariableName(queryContinuation.Identifier.ValueText);
+                        processingState.AddStatementProcessingMethod(
+                            s => AddToBlockTop(
+                                 SyntaxFactory.LocalDeclarationStatement(
+                                    SyntaxFactory.VariableDeclaration(
+                                        VarNameIdentifier,
+                                        SyntaxFactory.SingletonSeparatedList(
+                                            SyntaxFactory.VariableDeclarator(
+                                                queryContinuation.Identifier,
+                                                argumentList: null,
+                                                SyntaxFactory.EqualsValueClause(selectClause.Expression))))), s));
+
+                        return TryProcessQueryBody(queryContinuation.Body, processingState);
+                    }
+                }
+
+                return false;
+            }
+
+            private bool TryProcessWhereClause(WhereClauseSyntax whereClause, ProcessingState processingState)
+            {
+                if (IsValidWhereClauseSymbol(whereClause))
+                {
+                    processingState.AddStatementProcessingMethod(
+                         s => SyntaxFactory.Block(
+                             SyntaxFactory.IfStatement(
+                                 ProcessCondition((whereClause).Condition),
+                                 s)));
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool IsValidWhereClauseSymbol(WhereClauseSyntax whereClause)
+            {
+                if (!TryGetMethodSymbol(whereClause, out var methodSymbol))
+                {
+                    return false;
+                }
+
+                var returnType = methodSymbol.ReturnType;
+                if (returnType.Name != "IEnumerable") // TODO better that name check?
```

Comment:
you can do .OriginalDefinition and check if its' special type is System_Collections_Generic_IEnumerable_T